### ISSUE TITLE
stack prefers ghc from .stack-work\install\…\bin over the one from PATH

### ISSUE THREAD

**Issue by mwu-tow** [commercialhaskell/stack#1052]
I have a package that compiles to the executable named `ghc.exe`. 
When building first time, everything works fine. The install step places the built `ghc.exe` in the `.stack-work\install\x86_64-windows\lts-2.22\7.8.4\bin` directory. As expected.

However, any subsequent `stack build` command will fail:
```
    setup-Simple-Cabal-1.22.4.0-x86_64-windows-ghc-7.8.4.exe: The program 'ghc' version >=6.4 is required but the version of C:\z\.stack-work\install\x86_64-windows\lts-2.22\7.8.4\bin\ghc.exe could not be determined.
    setup-Simple-Cabal-1.22.4.0-x86_64-windows-ghc-7.8.4.exe: fd:5: hGetContents: invalid argument (invalid byte sequence)
```
To built it again I have to delete `.stack-work` directory.

It is because stack tried to use `.stack-work\install\x86_64-windows\lts-2.22\7.8.4\bin\ghc.exe` — that is not an actual GHC but the locally built package that was just named so. 

I expect that `stack` should prefer ghc.exe from PATH or its `ghc-paths` over the built executables. The results of build should not affect subsequent builds, and overwriting the compiler executable should not be implicit.

**Comment by snoyberg** [comment]
I've pushed a patch to master to pass in `--with-ghc` to the configure step, which should fix this. Can you give it a shot?

In general with build tools, however, we _want_ to use them for building things (e.g., a Stack-built alex should be used instead of whatever's on the PATH). I'm OK special-casing this for GHC, but in general having an executable shadow the name of another build tool should be considered an error.

**Comment by mwu-tow** [comment]
Actually I'm a little surprised that stack-built build tools should be preferred over the PATH. I'd expect that `stack build` simply builds things and should not affect how subsequent `stack build`. Doesn't it weaken the rule of reproducible builds?  If I wanted, I'd try using something like `stack exec stack build` (does this makes sense?) or just `install` the build tools. 
[Not saying you should change how things work, just thought it might me worth of noting.]

Is there any way I can prevent build artifacts from affecting the build? 

Anyway, with the latest master everything works fine for me. Many thanks! :-)

**Comment by snoyberg** [comment]
Consider the case of a tool like alex. Stack needs to build it in order for
other packages to use it, and we want to prefer that version of alex to
whatever happens to be on the PATH

**Comment by mwu-tow** [comment]
This fix seems to be broken in 0.1.6.0 release. Once again the compiled ghc gets used instead of the original one.

It works fine if I use your commit 0548ee2e0128f35b1d7ba5f5b7a28ef83b3fb861 but when I upgrade to release version it is broken again. Should I try checking when exactly it got broken?

**Comment by snoyberg** [comment]
That would be great.

**Comment by mwu-tow** [comment]
Done some bisecting, apparently the regression comes from 747cd0f289c51d31bbffb58341726284a622bee7 . 
Could it be that GHC used to compile `Setup.hs` is taken from the build?

**Comment by snoyberg** [comment]
Would you be interested in trying your hand at a PR based on the previous
fix for this?

**Comment by mwu-tow** [comment]
I can give it my try.

**Comment by mwu-tow** [comment]
I'm closing this issue, it is fixed by #1210 .
